### PR TITLE
[sdl] update to SDL3

### PR DESCRIPTION
**Pull request by akallabeth**
* port sdl client to upcoming `sdl3`, `sdl3_image` and `sdl3_ttf`
* add basic clipboard support
* make both, `SDL2` and `SDL3` client installable side by side or replacing each other (configurable)
* default to `OFF` for `SDL3` client build

### CODE REVIEW

**Comment by AkechiShiro** [comment]
Hi @akallabeth, thanks a lot for your work on this feature, is this port still on going as of right now ? It seems you are most likely busy working on other features at the moment ?

Is there any way someone that has no prior knowledge in this codebase help move this forward ? I have little experience with CMake (more with CMake than with C++)

**Comment by akallabeth** [comment]
@AkechiShiro well, waiting for `SDL3` to become a real thing (as in some release)
as you can see the port is not compatible with `SDL2` so and maintaining both will become quite a pain, so this is on hold until that time.

**Comment by AkechiShiro** [comment]
@akallabeth A prerelease was done for SDL3 recently was done : https://github.com/libsdl-org/SDL/releases/tag/prerelease-3.1.0

Just thought you might want to get this news !

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Comment by AkechiShiro** [comment]
Can this be used yet or tested in any way? Just would like to know if users could beta test and report issues on SDL3's preview version for `FreeRDP` ?

**Comment by akallabeth** [comment]
> Can this be used yet or tested in any way? Just would like to know if users could beta test and report issues on SDL3's preview version for `FreeRDP` ?

yes.
1. install `SDL`, `SDL_ttf` and `SDL_image` from their `git main branch` to some location (`/tmp/freerdp` if you don´t want a permanent location that is gone after reboot)
2. install `FreeRDP` to the same location (use `cmake -DCMAKE_INSTALL_PREFIX=/tmp/freerdp ...`
3. start `sdl3-freerdp` (that is the name of the `SDL3` client vs `sdl-freerdp` as the name of the `SDL2` client)

**Comment by AkechiShiro** [comment]
@LDprg @oskardotglobal any of you could do some testing with winapps with this experimental SDL3 backend ? If you cannot, I'll see if I can take some time to do so and report back my findings here so we can squash bugs in advance.

**Comment by akallabeth** [comment]
@hardening @mfleisz @bmiklautz @pnowack I´d appreciate a review (does not have to be thorough) of this pr.
as it only introduces a (new) client based on `SDL3` (with the required `CMake` glue and extended by clipboard, with at least text working) I consider merging this early and fix issues in follow up pull requests.

this way we can get (at least some) feedback from interested parties.

File: client/SDL3/dialogs/res/CMakeLists.txt.orig [hunk 1]
```diff
@@ -0,0 +1,99 @@
+
```
Comment (hardening) [inline client/SDL3/dialogs/res/CMakeLists.txt.orig:1]: This file should not be part of the PR ?

**Review by hardening — APPROVED** [review]
It seems like a good start. There's probably things that are common between the SDL2 and SDL3 client but this is something we can make common later.

File: client/SDL3/sdl_clip.cpp [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+	}
+	if (SDL_HasClipboardData(mime_html.c_str()))
+		clientFormatNames.push_back(type_HtmlFormat);
+
+	for (auto& mime : mime_bitmap)
+	{
+		if (SDL_HasClipboardData(mime.c_str()))
+		{
+			clientFormats.push_back({ CF_DIB, nullptr });
+			clientFormats.push_back({ CF_DIBV5, nullptr });
```
Comment (pnowack) [inline client/SDL3/sdl_clip.cpp:165]: If I remember correctly, the conversion from and to `CF_DIBV5` isn't implemented in `wClipboard`, so this will unlikely work (`CF_DIB` works though).

File: client/SDL3/sdl_clip.cpp [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+		clientFormats.push_back({ ClipboardRegisterFormat(_system, name.c_str()), name.data() });
+	}
+
+	std::sort(clientFormats.begin(), clientFormats.end(),
+	          [](const auto& a, const auto& b) { return a < b; });
+	auto u = std::unique(clientFormats.begin(), clientFormats.end());
+	clientFormats.erase(u, clientFormats.end());
+
+	const CLIPRDR_FORMAT_LIST formatList = {
+		.common = { .msgType = CB_FORMAT_LIST, .msgFlags = CB_RESPONSE_OK },
```
Comment (pnowack) [inline client/SDL3/sdl_clip.cpp:194]: `CB_RESPONSE_OK` is invalid here (it is only allowed in `*Response` PDUs). In my opinion, the clipboard PDU serialization handling (in `channels/cliprdr/client/cliprdr_main.c`) should fail sending any non-response PDUs, when the `CB_RESPONSE_OK` or `CB_RESPONSE_FAIL` flags are set (because that is not protocol-conform).
(Or have an assertion there that those flags are not used there (asserting correct API usage).)

File: client/SDL3/sdl_clip.cpp [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+
+	if (mime_is_image(mime))
+		return CF_DIB;
+	if (mime_is_text(mime))
+		return CF_UNICODETEXT;
+
+	return 0;
+}
+
+UINT sdlClip::RecieveServerCapabilities(CliprdrClientContext* context,
```
Comment (pnowack) [inline client/SDL3/sdl_clip.cpp:326]: Typo: `RecieveServerCapabilities` -> `ReceiveServerCapabilities`

File: client/SDL3/sdl_clip.hpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	std::string getServerFormat(uint32_t id);
+	uint32_t serverIdForMime(const std::string& mime);
+
+  private:
+	static UINT MonitorReady(CliprdrClientContext* context,
+	                         const CLIPRDR_MONITOR_READY* monitorReady);
+
+	static UINT RecieveServerCapabilities(CliprdrClientContext* context,
```
Comment (pnowack) [inline client/SDL3/sdl_clip.hpp:113]: Typo: `RecieveServerCapabilities` -> `ReceiveServerCapabilities`

File: client/SDL3/dialogs/res/convert_res_to_c.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
+/**
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply client/SDL3/dialogs/res/convert_res_to_c.cpp:1]: @hardening did not yet have the time to split it out, but yes.

File: client/SDL3/sdl_clip.cpp (hunk 3 above)
Comment (akallabeth) [reply client/SDL3/sdl_clip.cpp:194]: not used in `cliprdr_packet_format_list_new` so no protocol violation.
can print a warning though, that might help spot issues in code.

File: client/SDL3/sdl_clip.cpp (hunk 2 above)
Comment (akallabeth) [reply client/SDL3/sdl_clip.cpp:165]: it is handled like that for other clients as well.

looks like the format is unused by windows nowadays?

File: client/SDL3/dialogs/res/CMakeLists.txt.orig (hunk 1 above)
Comment (akallabeth) [reply client/SDL3/dialogs/res/CMakeLists.txt.orig]: removed.

File: client/SDL3/sdl_clip.cpp (hunk 2 above)
Comment (akallabeth) [reply client/SDL3/sdl_clip.cpp:165]: I´ll keep it this way and (if need arises) we´ll implement the format

File: client/SDL3/sdl_clip.cpp (hunk 3 above)
Comment (akallabeth) [reply client/SDL3/sdl_clip.cpp]: ok, added a warning with 21456dab097d003b2236e164d73fe97c7e1f3923

**Comment by akallabeth** [comment]
@freerdp-bot test again please...
